### ISSUE TITLE
Feature: Ability to add anchor and internal links for Footnotes and Table of Contents

### ISSUE THREAD

**Issue by alundgard** [projectblacklight/spotlight#3499]
_Content copied over from https://github.com/sul-dlss/exhibits/issues/1979 and the Spotlight Community Roadmap._

#### Footnotes user story

As an exhibit creator who is publishing exhibits with scholarly essays, I want to be able to add footnotes in the essay that link from the footnote indicator to the footnote itself and back. I would also like lengthy essays to have a table of contents that has anchor links for the page. This is so that scholarly essays can include proper references, and readers can navigate easily across headings, the footnotes, and essay text.

We are producing more Spotlight exhibits with scholarly content. Currently, we don’t have a good way to have readers be able to easily jump to a footnote and back to the running text, or between sections of the essay. Many new exhibits with scholarly essays are being planned.

See current workaround examples (all for footnotes):

- https://exhibits.stanford.edu/watkins 
- https://exhibits.stanford.edu/seaside/feature/david-starr-jordan-m-d-ph-d-ll-d
- https://exhibits.stanford.edu/sunset/feature/eras-1898-1914

#### Accessibility use case

Stanford note: Anchor links are also important for accessibility. Use case: when several images on a feature page contain extensive text, the text must be transcribed but it is too long to include as alt text (because it is bad practice to have lengthy alt text way in excess of 250 characters). Our accessibility solution is to create an additional feature page, and put all transcribed text on the page. But, the text for each image must be separated by a heading indicating what the image is. We can put a link in the widget to the page with transcriptions, but we should be able to link to the heading on the page for the transcribed text for each image.

### Feature details & acceptance criteria

- Ability to add anchor links as `id` attributes on various HTML elements (likely including headers, within-paragraph text spans, and list items).
- Ability to add anchor links within text paragraphs that link to footnotes on the page, and then link back to the text paragraph position.
- Ability to add anchor links within a list to create a Table of Contents linking to section headings on the page.
- Ability to add anchor links within image widget text paragraphs that link to specific headings appearing on a separate feature page.

### Other notes & comments
- Could be resolved by implementing TinyMCE WYSIWYG as a Sir Trevor widget.
- NLM note: Should this feature request be separated into two?  NLM provides a link to OCR/transcripts from a feature page to a new page to display the item.  See "Download" button at https://profiles.nlm.nih.gov/spotlight/kk/catalog/nlm:nlmuid-101584571X121-doc

**Comment by ebtoner** [comment]
Here is some thoughts on potential solutions for this issue:

I focused my initial investigation on what is recommended / built out for Markdown, since the text-based widgets in Spotlight rely on this syntax for advanced formatting. Basic Markdown syntax does not include anchor links, but many Markdown processors have built out extensions related to table of contents and footnotes. I looked mostly at [The Markdown Guide](https://www.markdownguide.org/) for general documentation and at the Ruby gem that we currently use for Markdown processing, [RedCarpet](https://github.com/vmg/redcarpet). (Note that we also use the Ruby gem [Github Markup](https://github.com/github/markup), but all that seems to do in our configuration is to call the RedCarpet renderer.) 

**Table of Contents and Heading IDs**

Some Markdown processors, including RedCarpet, include extensions to automatically add ids to all of their user-generated headings. Using RedCarpet’s extension, when you render the Markdown heading `## Spotlight Test`, the html output is `<h2 id="spotlight-test”>Spotlight Test</h2>`.

A simple solution for table of contents would be to enable this extension. Spotlight already allows for the use of relative links (though this could be better documented in the app), so once heading ids are included, users could manually build out table of contents with links to section headings on the same or different pages.

The advantages of this approach are: 1) It’s simple to implement. 2) There is no user input required to anchor the headings. The disadvantages of this approach are: 1) Longer titles would yield unwieldy ids. 2) If two headings on a page have the same title, there is no mechanism currently in place to prevent the reuse of the same id. 

Alternatively, we could explore this [extended syntax](https://www.markdownguide.org/extended-syntax/#heading-ids), highlighted by The Markdown Guide. However, I haven't been able to find this specific syntax implemented in any of the Markdown processors that I reviewed, so we would likely have implement it ourselves.

**Footnotes**

Many Markdown processors, including RedCarpet, also include extensions to add footnotes. Unlike the heading ids, though, this is not easily configured in Spotlight. Markdown processors generally assume that all of your content is in a single document. Because of this, RedCarpet’s footnotes extension will not work unless the inline footnote references and footnote definitions at the end of the page are processed together. In Spotlight, we render the Markdown syntax within each widget, so we would need to make some significant changes to our rendering process to make this work (probably not a good idea, I'm assuming).

One possible alternative would be to create a new “Footnotes” widget, based on the common [Markdown syntax for footnotes](https://www.markdownguide.org/extended-syntax/#footnotes). We could modify the existing list widget to create a footnotes definition list with appropriate markup and allow for the use of recommended Markdown inline formatting for footnote references (ex. “This is a text with a footnote[^1]”) in our text-based widgets. This could get messy, though, so I’m not sure if it’s worth it or not. 

Here is what this could look like: 

_Edit page_
<img width="1304" height="451" alt="Image" src="https://github.com/user-attachments/assets/4145ddb8-9359-4726-9a3c-1e321233f698" />
_Display_
<img width="733" height="124" alt="Image" src="https://github.com/user-attachments/assets/e049e05e-77fd-450b-8dbf-e57576d4d9f5" />


**Other options**

We could explore broader TinyMCE WYSIWYG editor options, as already mentioned in the description, or allow for the input of raw HTML, like anchor tags. However, I wanted to explore the solutions with the narrower scope before heading in that direction.
